File: headless/src/main/org/nlogo/prim/_fasthatch.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.agent.AgentSet;
import org.nlogo.agent.Turtle;
import org.nlogo.api.LogoException;
import org.nlogo.api.Syntax;
import org.nlogo.nvm.Command;

// replaces _hatch when initialization block is empty

public final strictfp class _fasthatch
    extends Command {
  private final String breedName;

  public _fasthatch(String breedName) {
    this.breedName = breedName;
  }

  @Override
  public Syntax syntax() {
    return Syntax.commandSyntax
        (new int[]{Syntax.NumberType()},
            "-T--", true);
  }

  @Override
  public String toString() {
    return super.toString() + ":" + breedName;
  }

  @Override
  public void perform(final org.nlogo.nvm.Context context) throws LogoException {
    int numberOfTurtles = argEvalIntValue(context, 0);
    if (numberOfTurtles > 0) {
      Turtle parent = (Turtle) context.agent;
      if (breedName == _hatch.NO_BREED) {
        for (int i = 0; i < numberOfTurtles; i++) {
          workspace.joinForeverButtons(parent.hatch());
        }
      } else {
        AgentSet breed = world.getBreed(breedName);
        for (int i = 0; i < numberOfTurtles; i++) {
          Turtle child = parent.hatch();
          child.setBreed(breed);
          workspace.joinForeverButtons(child);
        }
      }
    }
    context.ip = next;
  }
}

File: headless/src/main/org/nlogo/prim/_fasthatch.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.api.Syntax
import org.nlogo.agent.Turtle
import org.nlogo.nvm.{ Command, Context }

// replaces _hatch when initialization block is empty

class _fasthatch(breedName: String) extends Command {

  override def syntax =
    Syntax.commandSyntax(
      Array(Syntax.NumberType), "-T--", true)

  override def toString =
    super.toString + ":" + breedName

  override def perform(context: Context) = {
    val count = argEvalIntValue(context, 0)
    if (count > 0) {
      val parent = context.agent.asInstanceOf[Turtle]
      val breed =
        if (breedName.isEmpty) world.turtles
        else world.getBreed(breedName)
      var i = 0
      while(i < count) {
        workspace.joinForeverButtons(
          parent.hatch(breed))
        i += 1
      }
    }
    context.ip = next
  }

}
